tangram_bundler/__init__.py

```diff
@@ -1,100 +1,235 @@
 #!/usr/bin/env python
 
 import os, sys, glob, yaml, zipfile
+from urlparse import urljoin
 
 # Append uniform textures
 # - uniforms could reference a texture already loaded from textures: {} or could explicitly define a texture file
-def addUniformTextureDependency(file_list, basePath, yaml_file, styleName, uniformName):
+def addUniformTextureDependency(fileList, basePath, rootNode, styleName, uniformName):
     referenceTexturePath = ""
     explicitUniformTexturePath = ""
-    key = yaml_file['styles'][styleName]['shaders']['uniforms'][uniformName]
+    key = rootNode['styles'][styleName]['shaders']['uniforms'][uniformName]
 
     if isinstance(key, basestring):
-        if ('textures' in yaml_file) and (key in yaml_file['textures']):
-            referenceTexturePath = basePath + '/' + yaml_file['textures'][ yaml_file['styles'][styleName]['shaders']['uniforms'][uniformName] ]['url']
-        explicitUniformTexturePath = basePath + '/' + yaml_file['styles'][styleName]['shaders']['uniforms'][uniformName]
+        if ('textures' in rootNode) and (key in rootNode['textures']):
+            referenceTexturePath = os.path.relpath(rootNode['textures'][ rootNode['styles'][styleName]['shaders']['uniforms'][uniformName] ]['url'], basePath)
+        explicitUniformTexturePath = os.path.relpath(rootNode['styles'][styleName]['shaders']['uniforms'][uniformName], basePath)
 
     if (os.path.exists(explicitUniformTexturePath)):
-        file_list.append(os.path.normpath(explicitUniformTexturePath))
+        fileList.append(explicitUniformTexturePath)
     elif (os.path.exists(referenceTexturePath)):
-        file_list.append(os.path.normpath(referenceTexturePath))
-
-
-# Append yaml dependences in yaml_file ('import' files) to another yaml file (full_yaml_file)
-def addDependencies(file_list, filename):
-    print "Adding dependencies for file: ", filename
-    file_list.append(os.path.normpath(filename))
-    folder = os.path.dirname(filename)
-    yaml_file = yaml.safe_load(open(filename))
-
-    # TODO:
-    #   - Check if any url actually exist and is local before doing nothing
+        fileList.append(referenceTexturePath)
 
+# Fetch all asset dependencies in the constructed rootNode
+def fetchDependencies(fileList, rootNode, basePath):
```

Comment:
Does this logic account for absolute `http(s)://` paths in addition to relative ones?

Reply:
There should also be handling for cases where a texture URL field actually has a `global.` value -- in Tangram JS, the decision was to defer parsing of these until the entire scene is merged (since their final value can't be known until all imports are complete and merged), and then they are resolved against the base scene's path. I think these would have to just be omitted entirely from a server-side bundling step?